File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/context/SlotRenderContext.java
package me.devnatan.inventoryframework.context;

import me.devnatan.inventoryframework.BukkitViewer;
import me.devnatan.inventoryframework.RootView;
import me.devnatan.inventoryframework.Viewer;
import me.devnatan.inventoryframework.component.Component;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.UnknownNullability;

public class SlotRenderContext extends SlotContext implements IFSlotRenderContext {

    private final Player player;
    private final Viewer viewer;
    private final Component component;
    private ItemStack item;
    private boolean cancelled;
    private boolean changed;
    private boolean forceUpdate;

    @ApiStatus.Internal
    public SlotRenderContext(int slot, @NotNull IFRenderContext parent, @Nullable Viewer viewer, Component component) {
        super(slot, parent);
        this.viewer = viewer;
        this.player = viewer == null ? null : ((BukkitViewer) viewer).getPlayer();
        this.component = component;
    }

    @Override
    public final Component getComponent() {
        return component;
    }

    @Override
    public final @UnknownNullability Player getPlayer() {
        return player;
    }

    @Override
    public final Object getResult() {
        return item;
    }

    @Override
    public final ItemStack getItem() {
        return item;
    }

    public final void setItem(ItemStack item) {
        this.item = item;
        setChanged(true);
    }

    @Override
    public final boolean isCancelled() {
        return cancelled;
    }

    @Override
    public final void setCancelled(boolean cancelled) {
        this.cancelled = cancelled;
    }

    @Override
    public void clear() {
        setItem(null);
    }

    @Override
    public final boolean hasChanged() {
        return changed;
    }

    @Override
    public final void setChanged(boolean changed) {
        this.changed = changed;
    }

    @Override
    public boolean isForceUpdate() {
        return forceUpdate;
    }

    @Override
    public void setForceUpdate(boolean forceUpdate) {
        this.forceUpdate = forceUpdate;
    }

    @Override
    public final boolean isOnEntityContainer() {
        return getContainer().isEntityContainer();
    }

    @Override
    public final Viewer getViewer() {
        return viewer;
    }

    @Override
    public final void closeForPlayer() {
        getParent().closeForPlayer();
    }

    @Override
    public final void openForPlayer(@NotNull Class<? extends RootView> other) {
        getParent().openForPlayer(other);
    }

    @Override
    public final void openForPlayer(@NotNull Class<? extends RootView> other, Object initialData) {
        getParent().openForPlayer(other, initialData);
    }

    @Override
    public final void updateTitleForPlayer(@NotNull String title) {
        getParent().updateTitleForPlayer(title);
    }

    @Override
    public final void updateTitleForPlayer(@NotNull Object title) {
        getParent().updateTitleForPlayer(title);
    }

    @Override
    public final void resetTitleForPlayer() {
        getParent().resetTitleForPlayer();
    }
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/context/SlotRenderContext.kt
package me.devnatan.inventoryframework.context

import me.devnatan.inventoryframework.MinestomViewer
import me.devnatan.inventoryframework.RootView
import me.devnatan.inventoryframework.Viewer
import net.minestom.server.entity.Player
import net.minestom.server.item.ItemStack
import org.jetbrains.annotations.ApiStatus

class SlotRenderContext
    @ApiStatus.Internal
    constructor(slot: Int, parent: IFRenderContext, private val viewer: Viewer?) :
    SlotContext(slot, parent), IFSlotRenderContext {
        override val player: Player = (viewer as MinestomViewer).player

        override var item: ItemStack = ItemStack.AIR
        private var cancelled = false
        private var changed = false
        private var forceUpdate = false

        override fun getResult(): ItemStack {
            return item
        }

        override fun isCancelled(): Boolean {
            return cancelled
        }

        override fun setCancelled(cancelled: Boolean) {
            this.cancelled = cancelled
        }

        override fun clear() {
            item = ItemStack.AIR
        }

        override fun hasChanged(): Boolean {
            return changed
        }

        override fun setChanged(changed: Boolean) {
            this.changed = changed
        }

        override fun isForceUpdate(): Boolean {
            return forceUpdate
        }

        override fun setForceUpdate(forceUpdate: Boolean) {
            this.forceUpdate = forceUpdate
        }

        override fun isOnEntityContainer(): Boolean {
            return container.isEntityContainer
        }

        override fun getViewer(): Viewer? {
            return viewer
        }

        override fun closeForPlayer() {
            parent.closeForPlayer()
        }

        override fun openForPlayer(other: Class<out RootView?>) {
            parent.openForPlayer(other)
        }

        override fun openForPlayer(
            other: Class<out RootView?>,
            initialData: Any,
        ) {
            parent.openForPlayer(other, initialData)
        }

        override fun updateTitleForPlayer(title: String) {
            parent.updateTitleForPlayer(title)
        }

        override fun resetTitleForPlayer() {
            parent.resetTitleForPlayer()
        }
    }
